asserts/constraint.go

```diff
@@ -354,3 +354,111 @@ func (matcher altAttrMatcher) match(apath string, v interface{}, ctx *attrMatchi
 	}
 	return fmt.Errorf("no alternative%s matches: %v", apathDescr, firstErr)
 }
+
+// DeviceScopeConstraint specifies a constraint based on which brand
+// store, brand or model the device belongs to.
+type DeviceScopeConstraint struct {
+	Store []string
+	Brand []string
+	// Model is a list of precise "<brand>/<model>" constraints
+	Model []string
+}
+
+var (
+	validStoreID         = regexp.MustCompile("^[-A-Z0-9a-z_]+$")
+	validBrandSlashModel = regexp.MustCompile("^(" +
+		strings.Trim(validAccountID.String(), "^$") +
+		")/(" +
+		strings.Trim(validModel.String(), "^$") +
+		")$")
+	deviceScopeConstraints = map[string]*regexp.Regexp{
+		"on-store": validStoreID,
+		"on-brand": validAccountID,
+		// on-model constraints are of the form list of
+		// <brand>/<model> strings where <brand> are account
+		// IDs as they appear in the respective model assertion
+		"on-model": validBrandSlashModel,
+	}
+)
+
+func detectDeviceScopeConstraint(cMap map[string]interface{}) bool {
+	// for consistency and simplicity we support all of on-store,
+	// on-brand, and on-model to appear together. The interpretation
+	// layer will AND them as usual
+	for field := range deviceScopeConstraints {
+		if cMap[field] != nil {
+			return true
+		}
+	}
+	return false
+}
+
+// compileDeviceScopeConstraint compiles a DeviceScopeConstraint out of cMap,
+// it returns nil and no error if there are no on-store/on-brand/on-model
+// constraints in cMap
+func compileDeviceScopeConstraint(cMap map[string]interface{}, context string) (constr *DeviceScopeConstraint, err error) {
+	if !detectDeviceScopeConstraint(cMap) {
+		return nil, nil
+	}
+	// initial map size of 2: we expect usual cases to have just one of the
+	// constraints or rarely 2
+	deviceConstr := make(map[string][]string, 2)
+	for field, validRegexp := range deviceScopeConstraints {
+		vals, err := checkStringListInMap(cMap, field, fmt.Sprintf("%s in %s", field, context), validRegexp)
+		if err != nil {
+			return nil, err
+		}
+		deviceConstr[field] = vals
+	}
+
+	return &DeviceScopeConstraint{
+		Store: deviceConstr["on-store"],
+		Brand: deviceConstr["on-brand"],
+		Model: deviceConstr["on-model"],
+	}, nil
+}
+
+type DeviceScopeConstraintCheckOptions struct {
+	UseFriendlyStores bool
+}
+
+// Check tests whether the model and the optional store match the constraints.
+func (c *DeviceScopeConstraint) Check(model *Model, store *Store, opts *DeviceScopeConstraintCheckOptions) error {
+	if model == nil {
+		return fmt.Errorf("cannot match on-store/on-brand/on-model without model")
+	}
+	if store != nil && store.Store() != model.Store() {
+		return fmt.Errorf("store assertion and model store must match")
+	}
+	if opts == nil {
+		opts = &DeviceScopeConstraintCheckOptions{}
+	}
+	if len(c.Store) != 0 {
```

Comment:
Niptick: this could drop one level of nesting, i.e.
`if len(c.Store) != 0 && !strutil.ListContains(c.Store, model.Store()) {`

(same for brand and model checks below).

Reply:
(I realized after gettting to the end of this PR that this is pre-existing code, just moved around)